src/clara_specs/core.clj

```diff
@@ -8,46 +8,61 @@
 (s/def ::destructured (s/and vector?
                              (s/cat :form ::core/binding-form)))
 
-(s/def ::test (s/and vector?
-                     (s/cat :test #{:test}
-                            :expr (s/* any?))))
 
-(s/def ::accumulator (s/and vector?
-                            (s/cat :binding-var ::variable-name
-                                   :sep #{'<-}
-                                   :fact-type any?
-                                   :_ #{:from}
-                                   :fact-expr ::fact-expr)))
+(s/def ::test-expr (s/and vector?
+                       (s/cat :test #{:test}
+                              :expr (s/* any?))))
+
+(s/def ::constraint any?)
+
+(s/def ::condition (s/and vector?
+                          (s/cat :type any?
+                                 :constraints (s/* ::constraint))))
+
+
+(s/def ::accumulator-expr (s/and vector?
+                                 (s/cat :result-binding ::variable-name
+                                        :<- #{'<-}
+                                        :accumulator any?
+                                        :from #{:from}
+                                        :condition ::condition)))
 
 (s/def ::boolean-expr (s/and vector?
-                             (s/cat :condition #{:and :or :not}
-                                    :expression (s/+ (s/or :boolean-expr ::boolean-expr
-                                                           :fact-expr ::fact-expr))))) ; TODO what do these look like?
-
-(s/def ::fact-expr (s/and vector?
-                          (s/cat :binding (s/? (s/cat :binding-var ::variable-name
-                                                      :sep #{'<-}))
-                                 :fact-type any?
-                                 :destructured-fact (s/? ::destructured)
-                                 :expressions (s/* any?))))
-
-(s/def ::condition (s/or :accumulator ::accumulator
-                         :fact-expr ::fact-expr
-                         :boolean-expr ::boolean-expr
-                         :test ::test))
+                             (s/cat :op #{:and :or :not :exists}
+                                    :conditions (s/+ (s/or :boolean-expr ::boolean-expr
+                                                           :condition ::condition)))))
+
+(s/def ::fact-binding-expr (s/and vector?
+                            (s/cat :fact-binding (s/? (s/cat :binding-var ::variable-name
+                                                             :sep #{'<-}))
+                                   :condition ::condition)))
+
+(s/def ::expression (s/or :accumulator-expr ::accumulator-expr
```

Comment:
I believe part of the goal in this review is to mirror the terminology used within Clara already to describe the various pieces of the rules. Those can mostly all be found @ https://github.com/cerner/clara-rules/blob/0.16.1/src/main/clojure/clara/rules/schema.clj#L25-L85

I'll link to that rather than point out each place that I think this differs. I will mention here though that the fact binding is considered part of a condition and a LHS is comprised of a sequence of conditions.

I don't think the word "expression" is only used down towards the forms that make up the constraint that are actually expressions in the typical sense of the word. 

I think with the given link above it should be clearer though.

Reply:
> LHS is comprised of a sequence of conditions [...] I don't think the word "expression" is only used down towards the forms that make up the constraint that are actually expressions in the typical sense of the word.

I tried defining LHS as a sequence of conditions but wasn't able to implement it (things weren't getting parsed correctly). We can try to fix this but I'm not sure how. Might not be interpreting it correctly but I thought the "expression" construct used here echoes http://www.clara-rules.org/docs/expressions/:

> A query or rule left-hand side is a sequence of conditions, all of which must be satisfied for the rule or query to match. Each condition is one of the following:
> - A fact expression, which selects a fact based on some given criteria.
> - A boolean expression, which is simply a hierarchical and/or/not structure of other expressions.
> - An accumulator, which is mechanism to reason about collections of facts
> - A test, which is an arbitrary Clojure S-expression that can run predicate-style tests on variables bound earlier in the rule or query.
>

Reply:
I didn't realize that the docs disagreed with the terminology of the schema. That's a bit confusing. I prefer the schema terminology since that is what is used pervasively throughout the implementation.

> I tried defining LHS as a sequence of conditions but wasn't able to implement it
I'm not sure of the issue here. I was only commenting on naming, not the actual structure. I hvaen't looked into the issue you mention @ https://github.com/cursive-ide/clara-specs/pull/2#discussion_r159171709 above though, if that is what you are referring to.

Reply:
>  I was only commenting on naming, not the actual structure.

@mrrodriguez Thought that might be the case after I wrote the comment. :) Do I understand correctly now that you suggest `boolean-expr` should be named `boolean-condition`, but for the structure to remain the same?

Reply:
@alex-dixon Yes.
I think you already merged this though, so I guess you got it all figured out! Sorry I overlooked it.